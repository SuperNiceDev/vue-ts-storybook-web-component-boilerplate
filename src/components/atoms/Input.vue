<template>
  <div
    :class="{
      [$style.root]: true,
      [$style.filledOrFocused]: computedFilledOrFocused
    }"
  >
    <div :class="$style.placeholder">
      {{computedPlaceholder}}
    </div>
    <input
      :class="$style.input"
      type="text"
      @focus="onFocus"
      @input="onInput"
      @blur="onBlur"
      v-model="stateValue"
    >
  </div>
</template>


<script lang="ts">
import {
  Vue,
  Component,
  Prop,
  Watch,
} from 'vue-property-decorator';

const DEFAULT_PLACEHOLDER = 'default placeholder';

@Component({
  mounted() {
    this.stateValue = this.value;
  },
})
export default class Input extends Vue {
  private stateValue:string = '';

  private focused:boolean = false;

  @Prop({ default: '' }) value!: string;

  @Prop({ default: '' }) placeholder!: string;

  @Prop({ default: () => {} }) onvaluechange!: Function;

  @Watch('value')
  onValueChanged(val: string, oldVal: string) {
    this.stateValue = val;
  }

  onFocus():void {
    this.focused = true;
  }

  onBlur():void {
    this.focused = false;
  }

  onInput(newInputValue:string):void {
    this.$emit('onValueChange', this.stateValue);
    if (this.onvaluechange) {
      /* eslint-disable-next-line */
      // console.log('onInput() this.onvaluechange: ', this.onvaluechange);
      this.onvaluechange(this.stateValue);
    }
  }

  get computedFilledOrFocused():boolean {
    return !!(this.focused || this.value);
  }

  get computedPlaceholder():string {
    const test = this.placeholder || DEFAULT_PLACEHOLDER;
    return test;
    // return 'fff';
  }
}
</script>


<style lang="scss" module>
.root {
  position: relative;
  width: 100%;
  border: 1px solid grey;

  .input {
    position: relative;
    width: 100%;
    height: 100%;
    padding: 2px 6px;
    background-color: transparent;
    font-size: inherit;
    border: none;
    border-radius: 0;
    outline: none;
    outline: 0;
    box-shadow: none;
  }

  .placeholder {
    position: absolute;
    top: 0;
    bottom: 0;
    margin: auto 0;
    left: 6px;
    color: grey;
    line-height: 1.36;
    transition: all 300ms ease-in-out;
  }

  &.filledOrFocused {
    .placeholder {
      top: -100%;
      transform: scale(0.9);
      transform-origin: left center;
    }
  }
}
</style>
