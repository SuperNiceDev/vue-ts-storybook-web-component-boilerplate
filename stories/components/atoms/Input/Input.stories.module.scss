.inputWrapper {
  width: 300px;
}
