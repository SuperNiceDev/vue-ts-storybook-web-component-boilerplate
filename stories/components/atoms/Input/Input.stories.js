import { action } from '@storybook/addon-actions';
import {
  withKnobs,
  text,
} from '@storybook/addon-knobs';

import Input from '@src/components/atoms/Input.vue';
import globalStyles from '@stories/styles/globals.module.scss';
import styles from './Input.stories.module.scss';
import markdown from './docs.md';


export default {
  title: 'components|atoms/Input',
  decorators: [withKnobs],
  parameters: {
//     notes: `
// Input component:\n
// - info 1\n
// - info 2\n
// `,
    notes: { markdown },
  }
};


export const input = () => ({
  components: {
    Input,
  },
  props: {
    value: {
      default: text('value', 'test value'),
    },
    placeholder: {
      default: text('placeholder', 'test placeholder'),
    },
  },
  computed: {
    computedGlobalStyles() {
      return globalStyles;
    },
    computedStyles() {
      return styles;
    },
  },
  methods: {
    onInputValueChange: (key) => {
      action('onInputValueChange(): ')(key);
    },
  },
  template: `<div :class="computedGlobalStyles.storyRoot">
    <div :class="computedStyles.inputWrapper">
      <Input
        :value="value"
        :placeholder="placeholder"
        @onValueChange="onInputValueChange"
      />
    </div>
  </div>`,
});

// input.story = {
//   parameters: {
//     notes: { markdown },
//   }
// };