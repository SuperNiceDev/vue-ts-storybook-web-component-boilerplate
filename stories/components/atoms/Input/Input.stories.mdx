import { Meta, Preview, Story } from '@storybook/addon-docs/blocks'
import Input from '@src/components/atoms/Input.vue';

<Meta title="components/atoms/Input" component={Input} />

# Input Docs

Some MDX docs for the `Input` component.

<Preview>
  <Story name="Input.mdx">
    {{
      template: `<Input placeholder="Platzhalter" />`,
    }}
  </Story>
</Preview>