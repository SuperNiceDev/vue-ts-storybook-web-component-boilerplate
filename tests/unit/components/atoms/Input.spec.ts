import {
  shallowMount,
  mount,
} from '@vue/test-utils';

import Input from '@/components/atoms/Input.vue';

const value = 'test value';
const placeholder = 'test placeholder';

describe('Input.vue', () => {
  test('is a Vue instance', () => {
    const wrapper = mount(Input);
    expect(wrapper.isVueInstance()).toBeTruthy();
  });

  test('matches snapshot', () => {
    const wrapper = shallowMount(Input, {
      propsData: {
        value,
        placeholder,
      },
    });
    expect(wrapper).toMatchSnapshot();
  });
});
