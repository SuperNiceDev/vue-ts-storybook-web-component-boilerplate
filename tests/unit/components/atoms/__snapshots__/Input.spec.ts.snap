// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input.vue matches snapshot 1`] = `
<div class="root filledOrFocused">
  <div class="placeholder">
    test placeholder
  </div> <input type="text" class="input">
</div>
`;
